hibernate-rx-core/src/main/java/org/hibernate/rx/loader/RxOuterJoinLoader.java

```diff
@@ -0,0 +1,383 @@
+					}
+					return list;
+				});
+	}
+
+	private CompletionStage<List<?>> doRxQuery(
+			final SessionImplementor session,
+			final QueryParameters queryParameters,
+			final boolean returnProxies,
+			final ResultTransformer forcedResultTransformer) throws HibernateException {
+
+		final RowSelection selection = queryParameters.getRowSelection();
+		final int maxRows = LimitHelper.hasMaxRows( selection ) ?
+				selection.getMaxRows() :
+				Integer.MAX_VALUE;
+
+		final List<AfterLoadAction> afterLoadActions = new ArrayList<AfterLoadAction>();
+
+		return executeRxQueryStatement(
+				getSQLString(), queryParameters, false, afterLoadActions, session,
+				resultSet -> {
+					try {
+						return processResultSet(
+								resultSet,
+								queryParameters,
+								session,
+								returnProxies,
+								forcedResultTransformer,
+								maxRows,
+								afterLoadActions
+						);
+					}
+					catch (SQLException sqle) {
+						throw getFactory().getJdbcServices().getSqlExceptionHelper().convert(
+								sqle,
+								"could not load an entity batch: " + MessageHelper.infoString(
+										getEntityPersisters()[0],
+										queryParameters.getOptionalId(),
+										session.getFactory()
+								),
+								sql
+						);
+					}
+				}
+		);
+	}
+
+	private CollectionPersister[] collectionPersisters(Object resultSetId, SharedSessionContractImplementor session) {
+		final CollectionPersister[] collectionPersisters = getCollectionPersisters();
+		if ( collectionPersisters != null ) {
+			for ( CollectionPersister collectionPersister : collectionPersisters ) {
+				if ( collectionPersister.isArray() ) {
+					//for arrays, we should end the collection load before resolving
+					//the entities, since the actual array instances are not instantiated
+					//during loading
+					//TODO: or we could do this polymorphically, and have two
+					//      different operations implemented differently for arrays
+					endCollectionLoad( resultSetId, session, collectionPersister );
+				}
+			}
+		}
+		return collectionPersisters;
+	}
+
+	// Added this just to catch the exception and make the code more readable
+	private CompletionStage<Object> rxRows(ResultSet rs, QueryParameters queryParameters, SharedSessionContractImplementor session, boolean returnProxies, ResultTransformer forcedResultTransformer, EntityKey optionalObjectKey, LockMode[] lockModesArray, boolean createSubselects, List subselectResultKeys, List hydratedObjects, EntityKey[] keys) {
+		try {
+			return getRxRowFromResultSet( rs, session, queryParameters, lockModesArray, optionalObjectKey, hydratedObjects,
+					keys, returnProxies, forcedResultTransformer
+			);
+		} catch (SQLException sqle) {
+			throw getFactory().getJdbcServices().getSqlExceptionHelper().convert(
+					sqle,
+					"could not load an entity batch: " + MessageHelper.infoString(
+							getEntityPersisters()[0],
+							queryParameters.getOptionalId(),
+							session.getFactory()
+					),
+					sql
+			);
```

Comment:
We have a teensy bit of code duplication here. Let's factor this `try`/`catch` into a function. I can do it after you merge.

Reply:
Actually fixing this properly is sorta harder than it looks. Not worth the effort.